DOCS-#3747: Improve documentation of IO part (#3798)
modin-project/modin: docs/developer/architecture.rst

Data Ingress
''''''''''''

.. note::

>    Data ingress operation in Modin is operation of data import from data source to
>    partitions and vice versa for data egress operation, speed-up can be achieved
>    by performing operations in partitions in parallel.

Reviewer: ```suggestion
   Data ingress operations (e.g. ``read_csv``) in Modin load data from the source into
   partitions and vice versa for data egress (e.g. ``to_csv``) operation.
   Improved performance is achieved by reading/writing in partitions in parallel.
```

> Data ingress operations workflow diagram is shown below. After ingress function from pandas API
> is called, compiled query from pandas API is passed to the ``FactoryDispatcher``, which
> forwards query to execution spefic IO class (for Ray engine and pandas storage format IO class
> will be named ``PandasOnRayIO``). This class defines Modin frame and query
> compiler classes and ingress functions, which could be get from combination of the engine-specific
> class for managing remote tasks, class for data parsing on the workers by specific execution engine
> and file format (``PandasCSVParser`` for example) and class for files handling of concrete
> file format including chunking for execution on the head node (see dispatcher classes
> implementation :doc:`details </flow/modin/core/io/index>`). Using resulting IO class data
> is imported into Modin DataFrame for further processing.

Reviewer: ```suggestion
The data ingress operations workflow diagram is shown below. After a data 
ingress function from pandas API (e.g. ``read_csv``) is called, the compiled
query from pandas API is passed to the ``FactoryDispatcher``, which forwards
the query to execution specific IO class. For example, ``PandasOnRayIO`` is the
name of the IO class for cases using the Ray engine and pandas storage format. 
This IO class defines the Modin frame and query compiler classes to ensure correct
object is constructed. It additionally exposes supported
 ingress functions, which are mix-ins containing a combination of the engine-specific
class for deploying remote tasks, the class for defining how to parse the given
file format (``PandasCSVParser`` for example), and class that handles
file format specific chunking on the head node (see dispatcher classes
implementation :doc:`details </flow/modin/core/io/index>`). This ouput from
the IO class ingress function is formatted into a Modin DataFrame for further processing.
```

.. image:: /img/generic_data_ingress.svg
   :align: center

Data Egress
'''''''''''

> Data egress operation is similar to ingress operations up to execution-specific IO class
> functions construction. Egress function of this class are difined slightly different from
> ingress functions and created only specifically for engine since partitions already have
> information about it's storage format. Using this IO class data is exported from partitions
> to the target file.

Reviewer: ```suggestion
Data egress operations (e.g. ``to_csv``) are similar to ingress operations up to
execution-specific IO class functions construction. Egress functions of this class
are defined slightly different from ingress functions and created only
specifically for engine since partitions already have information about its storage
format. Using this IO class, data is exported from partitions to the target file.
```

.. image:: /img/generic_data_egress.svg
   :align: center

> 

Reviewer 2: I think we can take some description from "Factory Dispatcher" section below and combine with description of data ingress and data egress. "Factory Dispatcher" section might be dropped at all.

Author: Ok, "Factory Dispatcher" section was integrated into "Data Ingress" section.

Data Ingress
''''''''''''

.. note::
   Data ingress operations (e.g. ``read_csv``) in Modin load data from the source into
   partitions and vice versa for data egress (e.g. ``to_csv``) operation.
   Improved performance is achieved by reading/writing in partitions in parallel.

> After a data ingress function of the pandas API layer (e.g. ``read_csv``) is called, the user query is passed from that layer to the :doc:`Factory Dispatcher </flow/modin/core/execution/dispatching>`,
> which defines a factory specific for the execution, which in turn exposes the IO class
> whose responsibility is to perform a parallel read/write from/to a file. The resulting IO class contains IO methods with interfaces that correspond to pandas IO functions and its name represents the execution used. The IO class defines the Modin Dataframe and Query Compiler classes specific for the execution engine and storage format to ensure the correct object is constructed, as well as declares IO methods that are mix-ins containing a combination of the engine-specific class for deploying remote tasks, the class for parsing the given file format and the class handling the chunking of the format-specific file on the head node (see dispatcher classes implementation :doc:`details </flow/modin/core/io/index>`). The output from the IO class data ingress function is formatted into a Modin Dataframe.
> 

Reviewer: ```suggestion
Data ingress starts with a function in the pandas API layer (e.g. ``read_csv``). Then the user's 
query is passed to the :doc:`Factory Dispatcher </flow/modin/core/execution/dispatching>`,
which defines a factory specific for the execution. The factory for execution contains an IO class
(e.g. ``PandasOnRayIO``) whose responsibility is to perform a parallel read/write from/to a file.
This IO class contains class methods with interfaces and names that are similar to pandas IO functions
(e.g. ``PandasOnRayIO.read_csv``). The IO class declares the Modin Dataframe and Query Compiler
classes specific for the execution engine and storage format to ensure the correct object is constructed.
It also declares IO methods that are mix-ins containing a combination of the engine-specific class for
deploying remote tasks, the class for parsing the given file format and the class handling the chunking
of the format-specific file on the head node (see dispatcher classes implementation
:doc:`details </flow/modin/core/io/index>`). The output from the IO class data ingress function is
a :doc:`Modin Dataframe </flow/modin/core/dataframe/pandas/dataframe>`.
```

Data Ingress
''''''''''''

.. note::
   Data ingress operations (e.g. ``read_csv``) in Modin load data from the source into
   partitions and vice versa for data egress (e.g. ``to_csv``) operation.
   Improved performance is achieved by reading/writing in partitions in parallel.

The data ingress operations workflow diagram is shown below. After a data 
ingress function from pandas API (e.g. ``read_csv``) is called, the compiled

> query from pandas API is passed to the :doc:`Factory Dispatcher </flow/modin/core/execution/dispatching>`,

Reviewer 2: Inside flow/modin/core/execution/dispatching.rst there is the following sentence:
```
modin.pandas.read_csv calls FactoryDispatcher.read_csv, which calls .read_csv function of the factory...
```
Please change `.read_csv` to `._read_csv` function of the factory.

Author: Fixed.

> which provides IO class based on the selected execution. Resulting IO class will contain

Reviewer 2: FactoryDispatcher defines a factory specific for the execution. The factory, in turn, exposes the IO class whose responsibility is to perform a parallel read/write from/to a file. Look at pandas on Ray section as an example and correct this text.

Author: Corrected.

Data Ingress
''''''''''''

.. note::
   Data ingress operations (e.g. ``read_csv``) in Modin load data from the source into
   partitions and vice versa for data egress (e.g. ``to_csv``) operation.
   Improved performance is achieved by reading/writing in partitions in parallel.

> The data ingress operations workflow diagram is shown below. After a data 
> ingress function from pandas API (e.g. ``read_csv``) is called, the compiled
> query from pandas API is passed to the :doc:`Factory Dispatcher </flow/modin/core/execution/dispatching>`,
> which defines a factory specific for the execution, which in turn, exposes the IO class
> whose responsibility is to perform a parallel read/write from/to a file. Resulting IO class
> will contain IO methods with interfaces that correspond to pandas IO functions and it's name
> will represent used execution. For example, ``PandasOnRayIO`` is the
> name of the IO class for cases using the Ray engine and pandas storage format. 
> This IO class defines the Modin frame and query compiler classes to ensure correct
> object is constructed and IO methods are mix-ins containing a combination of the engine-specific
> class for deploying remote tasks, the class for defining how to parse the given
> file format (``PandasCSVParser`` for example), and class that handles
> file format specific chunking on the head node (see dispatcher classes
> implementation :doc:`details </flow/modin/core/io/index>`). This ouput from
> the IO class ingress function is formatted into a Modin DataFrame for further processing.

Reviewer 2: ```suggestion
After a data ingress function of the pandas API layer (e.g. ``read_csv``) is called, the user query is passed from that layer to the :doc:`Factory Dispatcher </flow/modin/core/execution/dispatching>`,
which defines a factory specific for the execution, which in turn exposes the IO class
whose responsibility is to perform a parallel read/write from/to a file. The resulting IO class contains IO methods with interfaces that correspond to pandas IO functions and its name represents the execution used. The IO class defines the Modin Dataframe and Query Compiler classes specific for the execution engine and storage format to ensure the correct object is constructed, as well as declares IO methods that are mix-ins containing a combination of the engine-specific class for deploying remote tasks, the class for parsing the given file format and the class handling the chunking of the format-specific file on the head node (see dispatcher classes implementation :doc:`details </flow/modin/core/io/index>`). The output from the IO class data ingress function is formatted into a Modin Dataframe for further processing.
```
I would prefer not to refer to execution-specific classes here so removed those references.
Also, what do you mean by `further processing` in `The output from the IO class data ingress function is formatted into a Modin Dataframe for further processing.`? Should be elaborated a bit.

Author: "Further processing" means resulting DataFrame will be used further in the ETL pipeline, but this is really confusing, so it was dropped.

.. image:: /img/generic_data_ingress.svg
   :align: center

Data Egress
'''''''''''

> Data egress operations (e.g. ``to_csv``) are similar to ingress operations up to
> execution-specific IO class functions construction. Egress functions of this class
> are defined slightly different from ingress functions and created only
> specifically for engine since partitions already have information about its storage
> format. Using this IO class, data is exported from partitions to the target file.
> 

Reviewer 2: ```suggestion
Data egress operations (e.g. ``to_csv``) are similar to data ingress operations up to
execution-specific IO class functions construction. Data egress functions of the IO class
are defined slightly different from data ingress functions and created only
specifically for the engine since partitions already have information about its storage
format. Using the IO class, data is exported from partitions to the target file.

```
